src/shipit_uplift/shipit_uplift/coverage_by_changeset_impl.py

```diff
@@ -0,0 +1,112 @@
+# -*- coding: utf-8 -*-
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import requests
+import whatthepatch
+from shipit_uplift.coverage import coverage_service
+
+
+def get_coverage_builds(changeset):
+    '''
+    This function returns the last successful build before a given
+    changeset and the first successful coverage build after the same
+    changeset.
+    '''
+    r = requests.get('https://hg.mozilla.org/mozilla-central/json-rev/%s' % changeset)
+    push_id = int(r.json()['pushid'])
+
+    # In a span of 15 pushes, we hope we will find successful coverage builds.
+    r = requests.get('https://hg.mozilla.org/mozilla-central/json-pushes?startID=%s&endID=%s' % (push_id - 8, push_id + 7))
+    data = r.json()
+
+    before_changeset = None
+    after_changeset = None
+    pushes = data.items()
+    pushes_before = [(pushid, pushdata) for pushid, pushdata in pushes if int(pushid) < push_id]
+    pushes_after = [(pushid, pushdata) for pushid, pushdata in pushes if int(pushid) >= push_id]
+
+    # Find the last coverage build before the changeset.
+    for pushid, pushdata in sorted(pushes_before, reverse=True):
+        changeset = pushdata['changesets'][-1]
+        if coverage_service.has_coverage(changeset):
+            before_changeset = changeset
+            break
+
+    assert before_changeset is not None
+
+    # Find the first coverage build after the changeset.
+    for pushid, pushdata in sorted(pushes_after):
+        changeset = pushdata['changesets'][-1]
+        if coverage_service.has_coverage(changeset):
+            after_changeset = changeset
+            break
+
+    assert after_changeset is not None
+
+    return (before_changeset, after_changeset)
+
+
+def generate(changeset):
```

Comment:
generate does not give enough information as to what this function does. Could you please add a docstring?

The other function names I think can easily be understood.

Reply:
Done with ff27f107286b94c22d787bec1bc59d858807a947.